### ISSUE TITLE
Bugs: Transient errors lead to unhandled 409/404 when acquiring and releasing an Azure lease

### ISSUE THREAD

**Issue by rmt2021** [dotnet/orleans#7693]
### Describe the two bugs
We encounter unhandled exceptions thrown by [`Release`](https://github.com/dotnet/orleans/blob/3895e070ea4b9870440a5c44eed93262923b73a5/src/Azure/Orleans.Streaming.AzureStorage/Providers/Lease/AzureBlobLeaseProvider.cs#L94-L98) method when transient errors happen. Depending on the specific transient error, the exception is triggered by error code 409 (LeaseIdMismatchWithLeaseOperation) or 404 (BlobNotFound). The exception is thrown from the Azure SDK API `ReleaseAsync` inside `Release`:
```
private Task Release(string category, AcquiredLease acquiredLease)
{
    var leaseClient = GetBlobClient(category, acquiredLease.ResourceKey).GetBlobLeaseClient(acquiredLease.Token);
    return leaseClient.ReleaseAsync();
}
```
There are two bugs when transient errors happen in different places:
#### Transient error leading to 404
The `Acquire` method that calls SDK APIs would NOT directly fail when transient errors (e.g., the request is refused because server is busy) happen since it has a catch block to handle the errors and will return a result containing the lease ID, HTTP status code, and exception.
```
private async Task<AcquireLeaseResult> Acquire(string category, LeaseRequest leaseRequest) {
    try {
        ...
        return new AcquireLeaseResult(new AcquiredLease(...), ResponseCode.OK, null);
    } catch (RequestFailedException e) {
        ResponseCode statusCode;
        switch (e.Status) {
            case 404:
            case 409:
            case 412: statusCode = ResponseCode.LeaseNotAvailable; break;
            default: statusCode = ResponseCode.TransientFailure; break; }
        return new AcquireLeaseResult(new AcquiredLease(leaseRequest.ResourceKey), statusCode, e);}}
```
However, when we release the lease via the `Release` method later, a 404 exception will be thrown since we did not acquire that lease successfully due to transient errors, and `Release` only takes the lease ID without considering the status code and exception in the previous acquire operation.

#### Transient error leading to 409
When releasing the lease, transient errors (e.g., timeout or package dropped) may happen during the transmission of the response. `ReleaseAsync` SDK API may retry releasing the lease multiple times until the lease is successfully released. This retry mechanism can lead to a 409 error even if the lease has already been released. One concrete scenario could be:
1. `leaseClient.ReleaseAsync` sends an HTTP request to release the lease first;
2. The lease is released by this first try, but a timeout happens, so the response never reaches the client; 
3. The SDK API will retry the release operation due to the timeout;
4. The second try fails and encounters a 409 LeaseIdMismatchWithLeaseOperation error since the lease has already been released by the first request;
5. The Azure SDK throws this 409 exception to Orleans, and `Release` in Orleans would not catch it (there is no try-catch block in `Release`).

### To Reproduce
**404 scenario**: When the acquire requests sent in `Acquire` method fail to arrive at the remote, BlobNotFound error code will appear in the response of the release request.
<details>
<summary>Logs for transient error leading to 404</summary>
Azure.RequestFailedException : The specified blob does not exist.
RequestId:23489625-0890-454d-9c12-d2f6b9c11bc2
Time:2022-04-03T05:17:14.089Z
Status: 404 (The specified blob does not exist.)
ErrorCode: BlobNotFound

Content:
ErrorCode: BlobNotFound
Message: The specified blob does not exist.
RequestId:23489625-0890-454d-9c12-d2f6b9c11bc2
Time:2022-04-03T05:17:14.089Z

at Azure.Storage.Blobs.BlobRestClient.ReleaseLeaseAsync(String leaseId, Nullable`1 timeout, Nullable`1 ifModifiedSince, Nullable`1 ifUnmodifiedSince, String ifMatch, String ifNoneMatch, String ifTags, CancellationToken cancellationToken);
   at Azure.Storage.Blobs.Specialized.BlobLeaseClient.ReleaseInternal(RequestConditions conditions, Boolean async, CancellationToken cancellationToken);
   at Azure.Storage.Blobs.Specialized.BlobLeaseClient.ReleaseAsync(RequestConditions conditions, CancellationToken cancellationToken);
   at Orleans.LeaseProviders.AzureBlobLeaseProvider.Release(String category, AcquiredLease[] acquiredLeases) in C:\Users\xxx\orleans\src\Azure\Orleans.Streaming.AzureStorage\Providers\Lease\AzureBlobLeaseProvider.cs:line 91;
   at TestExtensions.Runners.GoldenPathLeaseProviderTestRunner.ProviderCanReleaseLeases() in C:\Users\xxx\orleans\test\TestInfrastructure\TestExtensions\Runners\GoldenPathLeaseProviderTestRunner.cs:line 50;
--- End of stack trace from previous location ---
</details>

**409 scenario**: When the first release request sent by `ReleaseAsync` in `Release` succeeds in remote but its response never reaches the client, LeaseIdMismatchWithLeaseOperation will appear in the response of the second release request, which results in an unhandled exception.
<details>
<summary>Logs for transient error leading to 409</summary>
Azure.RequestFailedException : The lease ID specified did not match the lease ID for the container or blob.
RequestId:05bfe79c-2fc6-4edd-9895-dd27ff1b9f42
Time:2022-02-25T23:44:27.782Z
Status: 409 (The lease ID specified did not match the lease ID for the container or blob.)
ErrorCode: LeaseIdMismatchWithLeaseOperation

Content:
ErrorCode: LeaseIdMismatchWithLeaseOperation
Message: The lease ID specified did not match the lease ID for the container or blob.
RequestId:05bfe79c-2fc6-4edd-9895-dd27ff1b9f42
Time:2022-02-25T23:44:27.782Z

at Azure.Storage.Blobs.BlobRestClient.ReleaseLeaseAsync(String leaseId, Nullable1 timeout, Nullable1 ifModifiedSince, Nullable1 ifUnmodifiedSince, String ifMatch, String ifNoneMatch, String ifTags, CancellationToken cancellationToken)
   at Azure.Storage.Blobs.Specialized.BlobLeaseClient.ReleaseInternal(RequestConditions conditions, Boolean async, CancellationToken cancellationToken)
   at Azure.Storage.Blobs.Specialized.BlobLeaseClient.ReleaseAsync(RequestConditions conditions, CancellationToken cancellationToken)
   at Orleans.LeaseProviders.AzureBlobLeaseProvider.Release(String category, AcquiredLease[] acquiredLeases) in C:\Users\xxx\orleans\src\Azure\Orleans.Streaming.AzureStorage\Providers\Lease\AzureBlobLeaseProvider.cs:line 91
   at TestExtensions.Runners.GoldenPathLeaseProviderTestRunner.ProviderCanReleaseLeases() in C:\Users\xxx\orleans\test\TestInfrastructure\TestExtensions\Runners\GoldenPathLeaseProviderTestRunner.cs:line 50
--- End of stack trace from previous location ---
</details>

### Expected behavior & Possible Solution
The catch block of `Release` should catch the RequestFailedException and check whether it is a 404 or 409 response. If it is, the `Release` may need a handler to swallow the exception and print some logs.

### Additional context
We are willing to send a PR to help on this issue if swallowing 404 and 409 and printing logs is considered better practice than simply missing the exception handler for `Release`.

**Comment by benjaminpetit** [comment]
Let me rephrase to be sure I understood. I see two issues here:
- One is you tried to take a lease, but it failed. Then you try to release the lease, even though it wasn't successful. We could argue that the API isn't great, because you should not be able to release a lease that was not successful.
I don't think we should change something for that.
- The other is caused by the auto retry policy of the Azure SDK. I think it should be fine to just log a warning when getting a 409 when releasing the lease.

**Comment by benjaminpetit** [comment]
According to the comments in the code, it should be `null` if `Aquire` failed:
``` csharp
    public class AcquireLeaseResult
    {
        /// <summary>
        /// Acquired lease, which will be null if acquire or renew operation failed.
        /// </summary>
        AcquiredLease AcquiredLease { get; }
        [...]
    }
```
So we may want to fix that, but the rest of the code will also need fixing since there is a lot of code space that assume that `AcquiredLease` is always not null...

The best fix would be to redesign the API, it should not be possible to call `Release` on a failed lease IMO
